packages/block-editor/src/store/selectors.js

```diff
@@ -757,39 +758,7 @@ export function getSelectedBlocksInitialCaretPosition( state ) {
  * @return {Array} Multi-selected block client IDs.
  */
 export const getSelectedBlockClientIds = createSelector(
-	( state ) => {
-		const { selectionStart, selectionEnd } = state.selection;
-
-		if ( ! selectionStart.clientId || ! selectionEnd.clientId ) {
-			return EMPTY_ARRAY;
-		}
-
-		if ( selectionStart.clientId === selectionEnd.clientId ) {
-			return [ selectionStart.clientId ];
-		}
-
-		// Retrieve root client ID to aid in retrieving relevant nested block
-		// order, being careful to allow the falsey empty string top-level root
-		// by explicitly testing against null.
-		const rootClientId = getBlockRootClientId(
-			state,
-			selectionStart.clientId
-		);
-
-		if ( rootClientId === null ) {
-			return EMPTY_ARRAY;
-		}
-
-		const blockOrder = getBlockOrder( state, rootClientId );
-		const startIndex = blockOrder.indexOf( selectionStart.clientId );
-		const endIndex = blockOrder.indexOf( selectionEnd.clientId );
-
-		if ( startIndex > endIndex ) {
-			return blockOrder.slice( endIndex, startIndex + 1 );
-		}
-
-		return blockOrder.slice( startIndex, endIndex + 1 );
-	},
+	getSelectedBlockClientIdsUnmemoized,
 	( state ) => [
 		state.blocks.order,
 		state.selection.selectionStart.clientId,
```

Comment:
I don't fully understand why unmemoizing `getSelectedBlockClientIds` in particular should be faster. You write:

>  `createSelector` needs to keep a cache of all previous selector calls by arguments and then constantly look up cache for new calls. So if you have a 1000 blocks, that’s a huge loop of checks

This would be true if the selector had a `clientId` argument. Then the cache would contain a long linked list of memoized result values for each `clientId` argument value. But for this selector, the cache is a rather small structure:
1. At the top there is a `WeakMap` keyed by `state.blocks.order`. It should have just one record most of the time, because the old `order` array value are forgotten when the reducer state changes.
2. The value of the `WeakMap<Order>` record is another `WeakMap`, keyed by a special `LEAF_KEY` object, indicating there are no more object-like dependant values.
3. The value of the `WeakMap<LeafKey>` record is a linked list that always has just one record, for the `[]` argument array, because the selector has no arguments.
4. This one `[]` record contains the memoized return value.
5. The `[]` record is cleared when the current `clientId` is different from the last `clientId`. In other words, `rememo` remembers only the result for the last `clientId` used.

To summarize, there is just one three-level `cache.get( order ).get( LEAF_KEY ).find( r => r.clientId === clientId )` lookup in structures that have only one member. For reference, look at `rememo` source and how we have the `isUniqueByDependants === false` case for string/number `clientId`.

So, I think that unmemoizing this particular selector is controversial: it returns an array that deserves to have stable references, and I don't see where the speedup would come from.

For other selectors that return boolean values, I think unmemoization makes sense. Especially the ones that have arguments. It it plausible that recalculating the result from scratch is faster than maintaining large memoization maps.

One other such selector that comes to mind is `canInsertBlockType`. I've been changing it in #58262.

Reply:
> Then the cache would contain a long linked list of memoized result values for each clientId argument value.

Yes, that's true